### PR TITLE
Form Component for RecommendationRequest plus tests

### LINKED ISSUE #57: Form Component for RecommendationRequest plus tests

### PR DESCRIPTION
**Pull request by kanav9063**
closes #57

Description

- Adds the entire form component for RR
- Merge b2 and fixtures before merging this 

Tests

- All tests passing

### CODE REVIEW

**Review by lodetrick — APPROVED** [review]
LGTM! I see that you included fixtures, so as long as that isn't a merge conflict it should be good. Once the RequestType dropdown is complete we can also add that in a later PR

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+                type="text"
+                {...register("id")}
+                value={initialContents.id}
+                disabled
+              />
+            </Form.Group>
+          </Col>
+        )}
+
+        {/* Requester Email */}
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:50]: No.  This should be replaced with a dropdown component that shows all users from the Users table that have the Professor role. 

The dropdown should show the professors name and email, and the value field should be the id of the selected professor.

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+              isInvalid={Boolean(errors.requesterEmail)}
+              {...register("requesterEmail", {
+                required: true,
+              })}
+            />
+            <Form.Control.Feedback type="invalid">
+              {errors.requesterEmail && "Requester Email is required. "}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:67]: No.  The requster id is set in the backend, so there is no need to enter it.  The currently logged in user's email and name are already known.

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+              isInvalid={Boolean(errors.requesterName)}
+              {...register("requesterName", {
+                required: true,
+              })}
+            />
+            <Form.Control.Feedback type="invalid">
+              {errors.requesterName && "Requester Name is required. "}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:86]: Unnecessary.   Requester id is stored in the backend.
```suggestion

```

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js (hunk 2 above)
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:67]:
```suggestion

```

Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:67]: Set requesterId in the backend.   That way you get access to both the name and email.  

```suggestion

```

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+              isInvalid={Boolean(errors.professorEmail)}
+              {...register("professorEmail", {
+                required: true,
+              })}
+            />
+            <Form.Control.Feedback type="invalid">
+              {errors.professorEmail && "Professor Email is required. "}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:105]: No.
```suggestion
    {/* A dropdown component should go here, populated with all users that 
        have the Professsor flag set in the Users table.  Pass only the id of the selected
         professor to the backend.  From that, we can get the name and email of the professor */}
```

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+              {...register("professorName", {
+                required: true,
+              })}
+            />
+            <Form.Control.Feedback type="invalid">
+              {errors.professorName && "Professor Name is required. "}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
+      </Row>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:125]:
```suggestion

```

Not needed... see above.

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                MS program (other than CS Dept BS/MS)
+              </option>
+              <option value="PhD program">PhD program</option>
+              <option value="Other">Other</option>
+            </Form.Select>
+            <Form.Control.Feedback type="invalid">
+              {errors.requestType?.message}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:158]:
```suggestion
   {/* Replace with Dropdown selector that shows all values in the RequestType table */}
```

No.  The whole point of having a RequestType database table is that the request types are NOT hard coded.   This should be a dropdown of values from that table.

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+              isInvalid={Boolean(errors.submissionDate)}
+              {...register("submissionDate", {
+                required: "Submission Date is required.",
+              })}
+            />
+            <Form.Control.Feedback type="invalid">
+              {errors.submissionDate?.message}
+            </Form.Control.Feedback>
+          </Form.Group>
+        </Col>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:194]: No.   The submission date is calculated automatically in the backend.  The end user doesn't calculate it.

File: frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+            <Form.Label htmlFor="completionDate">Completion Date</Form.Label>
+            <Form.Control
+              data-testid="RecommendationRequestForm-completionDate"
+              id="completionDate"
+              type="datetime-local"
+              isInvalid={Boolean(errors.completionDate)}
+              {...register("completionDate")}
+            />
+          </Form.Group>
+        </Col>
```
Comment (pconrad) [inline frontend/src/main/components/RecommendationRequest/RecommendationRequestForm.js:208]: No.  Completion date is entered automatically in the backend LATER when the professor actually marks the request as complete.

However, we probably SHOULD have a "neededByDate", and this selector could be repurposed for that.  That would need to be added to the fixtures as well as to the entity and database migration files.

File: frontend/src/tests/components/RecommendationRequest/RecommendationRequestForm.test.js [hunk 9]
```diff
@@ -0,0 +1,156 @@
+import { render, waitFor, fireEvent, screen } from "@testing-library/react";
+import RecommendationRequestForm from "main/components/RecommendationRequest/RecommendationRequestForm";
+import { recommendationRequestFixtures } from "fixtures/recommendationRequestFixtures";
+import { BrowserRouter as Router } from "react-router-dom";
+
```
Comment (pconrad) [inline frontend/src/tests/components/RecommendationRequest/RecommendationRequestForm.test.js:5]: Adjust test after adjusting form as described above.

**Review by pconrad — CHANGES_REQUESTED** [review]
This form needs some significant revisions.

**Review by pconrad — APPROVED** [review]
ooops... the comment I previously typed was put on the wrong PR.

**Comment by pconrad** [comment]
image

**Pull request closed without merging**